src/librustc_typeck/check/mod.rs

```diff
@@ -705,9 +705,14 @@ impl Inherited<'a, 'tcx> {
             span_bug!(obligation.cause.span, "escaping bound vars in predicate {:?}",
                       obligation);
         }
-        self.fulfillment_cx
-            .borrow_mut()
-            .register_predicate_obligation(self, obligation);
+        let _ = self.fulfillment_cx
+            .try_borrow_mut()
+            .map_err(|e| self.tcx.sess.delay_span_bug(obligation.cause.span, &format!(
```

Comment:
This seems clearly correct in terms of fixing the ICE, but... where was this borrowed previously...?

Reply:
Yes, something seems fishy here.

Reply:
OK, I see what is going on. When we are reporting errors, we offer a callback to "improve" those errors. That callback is invoked with the fulfillment context borrowed. The callback (in this case) is then invoking normalization.

Reply:
I think we should just change it so that the borrow isn't held in the callback

Reply:
Specifically we would alter these lines

https://github.com/rust-lang/rust/blob/d63b24ffcc48f44ef09e0369e6516d6f2dec3520/src/librustc_typeck/check/mod.rs#L3109-L3112

to 

```rust
        let result = self.fulfillment_cx.borrow_mut().select_where_possible(self);
        if let Err(mut errors) = result {
            mutate_fullfillment_errors(&mut errors);
            self.report_fulfillment_errors(&errors, self.inh.body_id, fallback_has_occurred);
        }
```